go/vt/vtgate/planbuilder/operators/phases.go

```diff
@@ -85,11 +91,45 @@ func (p Phase) shouldRun(s semantics.QuerySignature) bool {
 		return s.SubQueries
 	case dmlWithInput:
 		return s.DML
+
 	default:
 		return true
 	}
 }
 
+// shouldUseBlockJoins returns true if the query should be rewritten to use BlockJoin.
+// This is only done for two table queries with no aggregation,
+// and the tables are not info_schema tables or derived tables
+func (p Phase) shouldUseBlockJoins(ctx *plancontext.PlanningContext) bool {
+	if ctx.Statement == nil {
+		return false
+	}
+
+	// Below are cases unsupported by block join:
+
+	// No DMLs
+	_, isSel := ctx.Statement.(sqlparser.SelectStatement)
+	if !isSel {
+		return false
+	}
+
+	// we only use block-joins for two normal table queries with no aggregation
+	stepOne := !ctx.SemTable.QuerySignature.Aggregation && len(ctx.SemTable.Tables) == 2
+	if !stepOne {
+		return false
+	}
+
+	// no information schema and non-normal tables
+	for _, tableInfo := range ctx.SemTable.Tables {
+		rt, isNormalTable := tableInfo.(*semantics.RealTable)
+		if !isNormalTable || rt.IsInfSchema() {
+			return false
+		}
+	}
+
+	return ctx.VSchema.AreBlockJoinsEnabled() || ctx.IsCommentDirectiveSet(sqlparser.DirectiveAllowBlockJoin)
```

Comment:
nit: we should check for directive and flag value upfront before checking other conditions.